kvserver: track incoming raft message queue byte size

tbg opened the pull request:
Release note: None


Jira issue: CRDB-14768

tbg committed:
kvserver: track incoming raft message queue byte size

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/79378)
<!-- Reviewable:end -->

aayushshah15 committed:
kvserver: avoid races where replication changes can get interrupted

This commit adds a safeguard inside
`Replica.maybeLeaveAtomicChangeReplicasAndRemoveLearners()` to avoid removing
learner replicas _when we know_ that that learner replica is in the process of
receiving its initial snapshot (as indicated by an in-memory lock on log
truncations that we place while the snapshot is in-flight).

This change should considerably reduce the instances where `AdminRelocateRange`
calls are interrupted by the mergeQueue or the replicateQueue (and vice versa).

Fixes https://github.com/cockroachdb/cockroach/issues/57129
Relates to https://github.com/cockroachdb/cockroach/issues/79118

Release note: none

tbg committed:
kvserver: wip

tbg committed:
kvserver: downgrade spammy "skipping snapshot" message

tbg committed:
wip

tbg committed:
wip: use memory budget for raft recv q

tbg committed:
logging.tags

tbg committed:
maybeTimeout bug fix

tbg committed:
fix replicate queue remove learner path

tbg committed:
actually remove the call to removeLearner

tbg committed:
fix the requeuing when learner inflight

tbg committed:
make limiters actually log

tbg committed:
drop-on-mem-alloc-failed
